### PR TITLE
misc changes to support redhat and older releases

### PR DESCRIPTION
**Pull request by DenisMishin**

### CODE REVIEW

File: infra/gravity/config.go [hunk 1]
```diff
@@ -15,6 +17,52 @@ import (
 	"gopkg.in/go-playground/validator.v9"
 )
 
+// OS represents OS vendor/version
+type OS struct {
+	Vendor, Version string
+}
+
+// Split returns os vendor and version
```
Comment (r0mant) [inline infra/gravity/config.go]: Misplaced comment.

File: infra/gravity/config.go [hunk 2]
```diff
@@ -15,6 +17,52 @@ import (
...
+	split := strings.Split(str, ":")
+	if len(split) != 2 {
+		return trace.BadParameter("OS should be in format vendor:version, got %q", b)
+	}
+	os.Vendor = split[0]
+	os.Version = split[1]
+	return nil
+}
+
+// Key returns back to serialized form
```
Comment (r0mant) [inline infra/gravity/config.go]: Either misplaced comment or Key -> String?

File: infra/gravity/config.go [hunk 3]
```diff
@@ -15,6 +17,52 @@ import (
...
+func (os *OS) String() string {
+	return fmt.Sprintf("%s:%s", os.Vendor, os.Version)
+}
+
+type StorageDriver string
```
Comment (r0mant) [inline infra/gravity/config.go]: Godoc.

File: infra/gravity/config.go [hunk 4]
```diff
@@ -15,6 +17,52 @@ import (
...
+	switch name {
+	case constants.DeviceMapper, constants.Overlay, constants.Overlay2, constants.Loopback, constants.ManifestStorageDriver:
+		*drv = StorageDriver(name)
+		return nil
+	default:
+		return trace.BadParameter("unknown storage driver %s", name)
+	}
+}
+
+// Driver validates and returns driver name
```
Comment (r0mant) [inline infra/gravity/config.go]: Doesn't look like this method validates anything.

File: infra/gravity/node_commands.go [hunk 5]
```diff
@@ -91,6 +91,10 @@ type InstallParam struct {
 	CloudProvider string `json:"cloud_provider,omitempty"`
 	// StateDir is the directory where all gravity data will be stored on the node
 	StateDir string `json:"state_dir" validate:"required"`
+	// OSVendorVersion is operating system and optional version separated by ':'
```
Comment (r0mant) [inline infra/gravity/node_commands.go]: OSVendorVersion -> OSFlavor

File: infra/gravity/config.go [hunk 6]
```diff
@@ -15,6 +19,52 @@ import (
...
+
+type StorageDriver string
+
+func (drv *StorageDriver) UnmarshalJSON(b []byte) error {
+	name, err := strconv.Unquote(string(b))
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	switch name {
+	case constants.DeviceMapper, constants.Overlay, constants.Overlay2, constants.Loopback, constants.ManifestStorageDriver:
```
Comment (a-palchikov) [inline infra/gravity/config.go]: We don't have `loopback` as a constant and it cannot be specified as a value for `--storage-driver` atm - it is implicitly implied if `devicemapper` is used w/o a device

File: lib/ssh/time.go [hunk 7]
```diff
@@ -94,17 +93,16 @@ func timeInRange(values []interface{}) bool {
...
-		*ts, err = strconv.ParseFloat(strings.TrimRight(line, "\n"), 64)
-		if err != nil {
-			return trace.Wrap(err, line)
+		if err := scanner.Err(); err != nil {
+			return trace.ConvertSystemError(err)
 		}
-
-		io.Copy(ioutil.Discard, r)
-		return nil
+		return trace.BadParameter("did not read any lines")
```
Comment (a-palchikov) [inline lib/ssh/time.go]:
```go
trace.BadParameter("empty input")
```

File: infra/gravity/config.go [hunk 8]
```diff
@@ -92,26 +143,26 @@ func (config ProvisionerConfig) WithNodes(nodes uint) ProvisionerConfig {
...
+	cfg.StateDir = filepath.Join(cfg.StateDir, fmt.Sprintf("%s%s", os.Vendor, os.Version))
 
 	return cfg
 }
 
 // WithStorageDriver returns copy of config with specific storage driver
-func (config ProvisionerConfig) WithStorageDriver(storageDriver string) ProvisionerConfig {
+func (config ProvisionerConfig) WithStorageDriver(storageDriver StorageDriver) ProvisionerConfig {
 	cfg := config
-	if storageDriver == constants.Loopback {
```
Comment (a-palchikov) [inline infra/gravity/config.go]: I wonder why you changed this?

**Pull request merged**